### PR TITLE
Enable support for the Context Info (alt+q) action in Dart (WIP)

### PR DESCRIPTION
**Pull request by DaveShuckerow**
From everything I can find about supporting this action, the declarationRangeHandler and corresponding activation in XML is all that's needed to enable the action.

However, at 4bb344e the declaration range handler is never invoked.  Is there anything I'm missing?

cc @alexander-doroshko

### CODE REVIEW

**Comment by alexander-doroshko** [comment]
Debug `ShowContainerInfoHandler.invoke()`. Looks like it relies on `StructureViewBuilder` and can work only if `getCurrentEditorElement()` returns `PsiElement`. However, Dart's one returns `String`.

**Comment by DaveShuckerow** [comment]
That was a good tip.  I've modified that code to find the corresponding PsiElement and the action works now.

Ready for review.

**Comment by DaveShuckerow** [comment]
Also, a question: what infrastructure is available to test the new action and make sure it continues to work?

File: Dart/src/com/jetbrains/lang/dart/ide/structure/DartStructureViewModel.java [hunk 1]
```diff
@@ -64,14 +64,15 @@ public StructureViewTreeElement getRoot() {
...
     if (getEditor() == null) return null;
 
     final DartAnalysisServerService service = DartAnalysisServerService.getInstance(getPsiFile().getProject());
     final Outline outline = service.getOutline(getPsiFile().getVirtualFile());
     if (outline == null) return null;
 
-    final Outline result = findDeepestOutlineForOffset(getEditor().getCaretModel().getOffset(), outline);
-    return DartStructureViewElement.getValue(result);
+    final int offset = getEditor().getCaretModel().getOffset();
+    return getPsiFile().getOriginalElement().findElementAt(offset);
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/structure/DartStructureViewModel.java]: I'm afraid this breaks `Navigate from source` functionality of the Structure View. See `DartStructureViewElement.getValue(Outline)` and its usages.
Unfortunately, we don't have tests for this. You may reproduce the regression manually: `Cmd+F12` won't focus the current element in the tree. Or toggle `Autoscroll from source` in the Structure tool window, it won't autoscroll as you move the caret in the editor.

And unfortunately, I'm not sure it's easy to have both Context Info and Structure View working.

Comment (DaveShuckerow) [reply Dart/src/com/jetbrains/lang/dart/ide/structure/DartStructureViewModel.java]: I believe I've fixed it.  Please take another look.

**Comment by alexander-doroshko** [comment]
I'm getting this exception while editing a file with Structure tool window open.
```
java.lang.NullPointerException
	at java.util.Objects.requireNonNull(Objects.java:203)
	at com.jetbrains.lang.dart.ide.structure.DartStructureViewElement.<init>(DartStructureViewElement.java:75)
	at com.jetbrains.lang.dart.ide.structure.DartStructureViewElement.lambda$getChildren$0(DartStructureViewElement.java:96)
	at com.intellij.util.containers.ContainerUtil.map2Array(ContainerUtil.java:1062)
	at com.jetbrains.lang.dart.ide.structure.DartStructureViewElement.getChildren(DartStructureViewElement.java:95)
	at com.jetbrains.lang.dart.ide.structure.DartStructureViewModel$DartStructureViewRootElement.getChildrenBase(DartStructureViewModel.java:139)
	at com.intellij.ide.structureView.impl.common.PsiTreeElementBase.doGetChildren(PsiTreeElementBase.java:107)
```

**Comment by DaveShuckerow** [comment]
Can you point me in the direction of the file and/or a way to reproduce?

**Comment by alexander-doroshko** [comment]
I'm not sure all steps are needed. File:
```
foo(){}
bar(){}
```
Structure tool window visible, both 'Autoscroll' actions enabled.
Select `bar(){}` and press Delete.

**Comment by DaveShuckerow** [comment]
I see; I'm able to reproduce by deleting the contents of a file.

It's possible that there are no PSI elements inside of a file at the cursor offset, in which case, we won't find one.

**Comment by DaveShuckerow** [comment]
Please take another look.

**Comment by alexander-doroshko** [comment]
Thanks! I'll test it a bit more and merge next week.

**Comment by alexander-doroshko** [comment]
Sorry for delay with this PR. I noticed that some tricky cases don't work perfectly but it was difficult to discuss it here. Specifically, the logic to find `myValue` for `DartStructureViewElement` looked a bit hacky. Also, consequent `Alt+Q` worked strange for tests/test groups. I hope I managed to address the issues. Please have a look, here are the [changes in the Structure View](https://github.com/JetBrains/intellij-plugins/commit/8e062460d4961dec7982551ac7386612bb913835#diff-35900d987289a83af101d18a9f7c3501), and here's [Context Info](https://github.com/JetBrains/intellij-plugins/commit/c503a0bde506ebabb767b28d8f05d24142c54dfa#diff-35900d987289a83af101d18a9f7c3501). Could you review/test it (I need it to cherry-pick to the release branch)?

**Comment by DaveShuckerow** [comment]
Your changes look good to me.  Thanks for showing some of what the analysis service can do.

**Pull request closed without merging**